### PR TITLE
Feature/home

### PR DESCRIPTION
**Pull request by Rodrigo-ade**
# Summary
- Add new page `index`.
- Add new components: `ActionButton`, `ConnectModal`, `KeyPairModal`.
- Add new `stellar` service to obtain different values from Stellar.
- Add tests to all the content. 

# Details
- `index` will have the options to `Connect` and `Get new keys`.
- `ConnectModal` will be used to get logged in.
- `KeyPairModal` will be used to create new key pairs. 

# Evidence 

## index page:
image

## KeyPairModal:
image

## ConnectModal:
image

### CODE REVIEW

File: cypress/tests/index.spec.ts [hunk 1]
```diff
@@ -1,10 +1,83 @@
 /// <reference types="cypress" />
-describe("example to-do app", () => {
-  beforeEach(() => {
-    cy.visit("http://localhost:3000");
+
+const DEFAULT_URL = "http://localhost:3000";
```
Comment (Genebson) [inline cypress/tests/index.spec.ts:3]: This should come from an environment variable

File: cypress/tests/index.spec.ts [hunk 2]
```diff
@@ -1,10 +1,83 @@
...
+
+context("Index", () => {
+
+  describe('Index page', () => {
+    before(() => {
+      cy.visit(DEFAULT_URL);
+    });
+
+    it('"Generate Key" button should exist', () => {
+      cy.get('[data-cy="keys-button"]').should('exist');
```
Comment (Genebson) [inline cypress/tests/index.spec.ts:16]: This could  be an alias, we can declare this inside our `beforeEach`

File: cypress/tests/index.spec.ts [hunk 3]
```diff
@@ -1,10 +1,83 @@
...
+    });
+
+    it('"Connect with secret key" button should exist', () => {
+      cy.get('[data-cy="connect-button"]').should('exist');
```
Comment (Genebson) [inline cypress/tests/index.spec.ts:20]: This could be an alias as well.

File: cypress/tests/index.spec.ts [hunk 4]
```diff
@@ -1,10 +1,83 @@
...
+    })
+  });;
+
+  describe('Key Modal', () => {
+    it('"Generate Key" button should open modal', () => {
```
Comment (Genebson) [inline cypress/tests/index.spec.ts:25]: We start our `it` messages with `Should...`

File: cypress/tests/index.spec.ts [hunk 5]
```diff
@@ -1,10 +1,83 @@
...
+    it('Should be closed when close is clicked', () => {
+      cy.get('[data-cy="close-button"]').click();
+      cy.get('[data-cy="key-modal"]').should('not.exist');
+    });
   });
 
-  it("Should render the body", () => {
-    cy.get("#__next").should("be.visible")
+  describe('Connect Modal', () => {
+    it('"Connect" button should open modal', () => {
```
Comment (Genebson) [inline cypress/tests/index.spec.ts:56]: Same here.

File: cypress/tests/index.spec.ts [hunk 6]
```diff
@@ -1,10 +1,83 @@
...
+      cy.get('[data-cy="connect-button"]').click();
+      cy.get('[data-cy="connect-modal"]').should('be.visible');
```
Comment (Genebson) [inline cypress/tests/index.spec.ts:58]: This could be an alias as well.

File: cypress/tests/index.spec.ts [hunk 7]
```diff
@@ -1,10 +1,83 @@
...
+    });
+
+    it('Should show error with invalid secret key', () => {
+      cy.get('[data-cy="key-input"').type(INVALID_PRIVATE_KEY);
```
Comment (Genebson) [inline cypress/tests/index.spec.ts:62]: This could be an alias as well.

File: src/components/connectModal/ConnectModal.tsx [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import Router from 'next/router';
+import { isValidSecretKey, getPublicKey } from '@/services/stellar';
```
Comment (Genebson) [inline src/components/connectModal/ConnectModal.tsx:2]: Only pages know about dependencies, components shouldn't know about them, they must stay pure.

Ask me about this, please.

File: src/components/keyPairModal/KeyPairModal.tsx [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+export function KeyPairModal({ setShowKeys, privateKey, setPrivateKey, publicKey, setPublicKey }: KeyPairModalProps) {
+  const [copyKeysButtonText, setCopyKeysButtonText] = useState<'Copy Keys' | 'Copied!'>('Copy Keys');
+
+  function handleCopyKeys() {
+    navigator.clipboard.writeText(`Public Key: ${publicKey} Private Key: ${privateKey}`).then(() => {
+      setCopyKeysButtonText('Copied!');
+      setTimeout(() => {
+        setCopyKeysButtonText('Copy Keys');
+      }, 3000);
```
Comment (Genebson) [inline src/components/keyPairModal/KeyPairModal.tsx:20]: Save this magic number inside a variable, please.

File: src/components/keyPairModal/KeyPairModal.tsx [hunk 10]
```diff
@@ -0,0 +1,83 @@
+import { useState } from 'react';
+import { ActionButton } from '../actionButton/ActionButton';
+
+export interface KeyPairModalProps {
```
Comment (Genebson) [inline src/components/keyPairModal/KeyPairModal.tsx:4]: We declare interfaces with an `I` at the beginning

File: src/components/keyPairModal/KeyPairModal.tsx [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+                      <p className="mt-5 text-purple-200">Private Key:</p>
+                      <code className="break-all text-purple-500" data-cy="private-key">{privateKey}</code>
+                    </div>
+                  </div>
+                </div>
+              </div>
+            </div>
+            <div className="justify-between bg-purple-highlight	px-4 py-3 sm:flex sm:px-6">
+              <ActionButton title={copyKeysButtonText} handleClick={() => handleCopyKeys()} cyAttribute="copy-button" />
+              <ActionButton title="Close" handleClick={() => handleCloseModal()} cyAttribute="close-button"/>
```
Comment (Genebson) [inline src/components/keyPairModal/KeyPairModal.tsx:76]: Why do we've this cyAttribute here?

File: src/pages/index.tsx [hunk 12]
```diff
@@ -1,72 +1,61 @@
...
-      </main>
+const Home: NextPage = () => {
+  const [privateKey, setPrivateKey] = useState('');
+  const [publicKey, setPublicKey] = useState('');
+  const [showKeys, setShowKeys] = useState(false);
+  const [showConnectModal, setShowConnectModal] = useState(false);
+  const [isValidKey, setIsValidKey] = useState(true);
+
+  function handleGenerateKeys() {
+    const keys = getRandomKeyPair();
```
Comment (Genebson) [inline src/pages/index.tsx:18]: You can destructure here.

File: src/pages/index.tsx [hunk 13]
```diff
@@ -1,72 +1,61 @@
...
-            <Image src="/vercel.svg" alt="Vercel Logo" width={72} height={16} />
-          </span>
-        </a>
-      </footer>
+  return (
+    <div className="grid h-screen items-center text-center">
+      <h1 className="text-6xl text-slate-400">Stellar Wallet</h1>
+      <div>
+        <ActionButton title="Generate new keys" handleClick={() => handleGenerateKeys()} cyAttribute="keys-button" />
+        <ActionButton title="Connect with secret key" handleClick={() => handleLogIn()} cyAttribute="connect-button" />
```
Comment (Genebson) [inline src/pages/index.tsx:33]: Same here.

File: src/services/stellar.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  const keyPair = Keypair.random();
+  const publicKey = keyPair.publicKey();
+  const privateKey = keyPair.secret();
+  return { publicKey, privateKey };
+}
+
+export function isValidSecretKey(secretKey: string): boolean {
+  const isValidKey = StrKey.isValidEd25519SecretSeed(secretKey);
+  return isValidKey;
+}
```
Comment (Genebson) [inline src/services/stellar.ts:18]:
```suggestion
export function isValidSecretKey(secretKey: string): boolean {
  return StrKey.isValidEd25519SecretSeed(secretKey);
}
```

File: src/services/stellar.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+export function getPublicKey(privateKey: string): string {
+  const publicKey = Keypair.fromSecret(privateKey).publicKey();
+  return publicKey;
+}
```
Comment (Genebson) [inline src/services/stellar.ts:22]:
```suggestion
export function getPublicKey(privateKey: string): string {
  return Keypair.fromSecret(privateKey).publicKey();
}
```

File: src/components/actionButton/ActionButton.tsx [hunk 16]
```diff
@@ -0,0 +1,19 @@
+export interface ActionButtonProps {
```
Comment (Genebson) [inline src/components/actionButton/ActionButton.tsx:1]: Same here.

File: cypress/tests/index.spec.ts [hunk 17]
```diff
@@ -1,10 +1,92 @@
...
+        .should('not.be.empty')
+        .should('have.length.at.least', 50);
+    });
+
+    it('Should copy keys when copy button is clicked', () => {
+      cy.get('@getKeysButton').click();
+      cy.get('[data-cy="copy-button"]').click();
+
+      cy.window().its('navigator.clipboard').invoke('readText')
+        .should('have.length.above',100)
```
Comment (Genebson) [inline cypress/tests/index.spec.ts:47]: Is Prettier not working well?
```suggestion
        .should('have.length.above', 100)
```

File: src/components/actionButton/ActionButton.tsx [hunk 18]
```diff
@@ -0,0 +1,19 @@
+export interface IActionButtonProps {
+  title: string;
+  handleClick: () => void;
+  cyAttribute?: string;
+}
+
+export function ActionButton({ title, handleClick, cyAttribute }: IActionButtonProps) {
+  return (
+    <button
+      data-cy={ cyAttribute ? cyAttribute : '' }
```
Comment (Genebson) [inline src/components/actionButton/ActionButton.tsx:10]: Is Prettier not working here as well?

I think we can avoid this by adding a custom `className` to get the element from our tests.

File: src/components/connectModal/ConnectModal.tsx [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+  isValidKey,
+  setIsValidKey,
+  privateKey,
+  setPrivateKey,
+  isValidSecretKey,
+  getPublicKey,
+  redirectToDashboard,
+}: IConnectModalProps) {
+  function handleConnect() {
+    if (!isValidSecretKey(privateKey)) {
```
Comment (Genebson) [inline src/components/connectModal/ConnectModal.tsx:25]:
```suggestion
    if (!isSecretKeyValid(privateKey)) {
```

File: src/components/connectModal/ConnectModal.tsx [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+      setIsValidKey(false);
```
Comment (Genebson) [inline src/components/connectModal/ConnectModal.tsx:26]:
```suggestion
      setIsKeyValid(false);
```

File: cypress/tests/index.spec.ts (hunk 17 above)
Comment (Rodrigo-ade) [reply cypress/tests/index.spec.ts:47]: Sorry, i forgot to run our formatter.

File: src/components/actionButton/ActionButton.tsx [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+  handleClick: () => void;
+}
+
+export function ActionButton({ title, handleClick }: IActionButtonProps) {
+  return (
+    <button
+      className={`action-button m-2 rounded border-2 border-solid border-gray-900 bg-transparent px-6 py-3 font-bold text-gray-300 hover:border-gray-600`}
+      onClick={() => {
+        handleClick();
+      }}
```
Comment (Genebson) [inline src/components/actionButton/ActionButton.tsx:12]: I think you can do this, let me know if it works.
```suggestion
      onClick={handleClick}
```

File: src/components/connectModal/ConnectModal.tsx [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+                    Invalid secret key
+                  </p>
+                ) : (
+                  ''
+                )}
+              </div>
+            </form>
+            <div className="justify-between bg-purple-highlight	px-4 py-3 sm:flex sm:px-6">
+              <ActionButton title="Connect" handleClick={() => handleConnect()} />
+              <ActionButton title="Close" handleClick={() => handleCloseConnectModal()} />
```
Comment (Genebson) [inline src/components/connectModal/ConnectModal.tsx:73]: Same here.

File: src/components/keyPairModal/KeyPairModal.tsx [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+                        {privateKey}
+                      </code>
+                    </div>
+                  </div>
+                </div>
+              </div>
+            </div>
+            <div className="justify-between bg-purple-highlight	px-4 py-3 sm:flex sm:px-6">
+              <ActionButton title={copyKeysButtonText} handleClick={() => handleCopyKeys()}/>
+              <ActionButton title="Close" handleClick={() => handleCloseKeyPairModal()} />
```
Comment (Genebson) [inline src/components/keyPairModal/KeyPairModal.tsx:83]: Same here.

**Review by Genebson — APPROVED** [review]

File: src/components/actionButton/ActionButton.tsx (hunk 21 above)
Comment (Rodrigo-ade) [reply src/components/actionButton/ActionButton.tsx]: It works, changes done.

**Pull request merged**